apps/prairielearn/src/middlewares/clientFingerprint.ts

```diff
@@ -0,0 +1,42 @@
+import asyncHandler from 'express-async-handler';
+import * as sqldb from '@prairielearn/postgres';
+
+const sql = sqldb.loadSqlEquiv(__filename);
+
+export default asyncHandler(async (req, res, next) => {
+  const params = {
+    ip_address: req.ip,
+    user_id: res.locals.user.user_id,
+    session_id: req.session.id,
+    user_agent: req.headers['user-agent'],
+    accept_language: req.headers['accept-language'],
+    accept: req.headers['accept'],
+  };
+  // Need a better way to find or insert a fingerprint. Currently doing this through two calls but would probably be better to do it in one.
+  // First call, check if fingerprint exists in the database
+  let client_fingerprint_id: number | null =
+    parseInt(
+      (await sqldb.queryZeroOrOneRowAsync(sql.select_client_fingerprint, params)).rows[0]?.id,
+    ) ?? null;
+  //if the fingerprint doesn't exist, create a new fingerprint
+  if (!client_fingerprint_id) {
+    client_fingerprint_id = (await sqldb.queryOneRowAsync(sql.insert_client_fingerprint, params))
+      .rows[0].id;
+  }
+
+  // Check if the client fingerprint matches the last fingerprint on the assessment id. If not, update the assessment instance with the new fingerprint and increment the fingerprint count.
+  if (
+    res.locals.assessment_instance &&
+    res.locals.assessment_instance?.last_client_fingerprint_id !== client_fingerprint_id
+  ) {
+    await sqldb.queryAsync(sql.update_assessment_instance_fingerprint, {
+      client_fingerprint_id,
+      assessment_instance_id: res.locals.assessment_instance?.id,
+      client_fingerprint_id_change_count:
+        res.locals.assessment_instance.client_fingerprint_id_change_count + 1,
+    });
+  }
```

Comment:
This is going to be susceptible to the issue I talked about last week where the count will increase when an instructor or other user views the student's assessment. I don't believe we want that to count towards the change count.

This would be avoidable if we computed the change count on the fly, though I understand why that's not a desirable approach.

My current thinking is that we only increment this count when the current user is either the owner of the assessment instance, or a member of the group that owns the assessment instance. I believe we'd want to check the authenticated user, not the authorized user. I can talk about the distinction there once we decide on an approach!

I'll let @mwest1066 chime in with thoughts here too.

Reply:
I think we want to update the change count on every access, rather than computing it later from the saved fingerprint data, so that we can cheaply display the change count for all assessment instances. I agree with @nwalters512's approach of only modifying the fingerprint fields for an assessment instance (both `last_client_fingerprint_id` and `client_fingerprint_id_change_count`) when the current _authenticated_ user is an owner (possibly via group membership) of the assessment instance. I also agree with @jonatanschroeder's approach below of updating both variables in a single SQL query.

Reply:
Per our meeting yesterday, we are now only updating the assessment instance fingerprint fields when the authenticated user is accessing the assessment instance. Also, we have updated SQL query to calculate the `client_fingerprint_id_change_count` to eliminate race conditions.

Reply:
I continued discussion of this in a new comment: https://github.com/PrairieLearn/PrairieLearn/pull/8798#discussion_r1409736082